### PR TITLE
TT-541 find mismatches contract headers

### PR DESCRIPTION
**Pull request by committedteadrinker**
## Description, Motivation and Context
This change is part of the mismatcher contract testing. This functionality is needed to verify and find mismatches on request and response header with the contract that's passed into spot.
https://airtasker.atlassian.net/browse/TT-541

## Checklist:

- [x] I've added/updated tests to cover my changes
- [x] I've created an issue associated with this PR

### CODE REVIEW

**Review by lfportal — CHANGES_REQUESTED** [review]

**Review by alephao — COMMENTED** [review]
It looks like the `Error` case of the `Result` type is not very useful for most cases since you're always transforming an `Error` into a `Mismatch` in the end.

I believe that the `Result` type might not be the best abstraction for our use case, in the future would be good to explore other ideas like the `Validated` type

File: lib/src/neu/verifications/contract-mismatcher.ts [hunk 1]
```diff
@@ -60,6 +82,165 @@ export class ContractMismatcher {
...
+          accumulator.push(...result.unwrap());
+        }
+        return accumulator;
+      },
+      []
+    );
+    return ok(mismatches);
+  }
+
+  private findMismatchOnResponseHeader(
```
[the remark this answers is not in the document]
Comment (committedteadrinker) [reply lib/src/neu/verifications/contract-mismatcher.ts:150]: Like the above, the endpoint path and method are needed for the error messages, I can unbox the object and pass in as 3 separate parameters if you want.

File: lib/src/neu/verifications/contract-mismatcher.ts [hunk 2]
```diff
@@ -60,6 +82,165 @@ export class ContractMismatcher {
...
+    return ok(mismatches);
+  }
+
+  private getResponseHeadersOnContractEndpoint(
+    endpoint: Endpoint,
+    userInputResponse: UserInputResponse
+  ): Result<Header[], Error> {
+    if (endpoint.responses.length > 0) {
+      for (const contractResponse of endpoint.responses) {
+        if (contractResponse.status === userInputResponse.statusCode) {
```
[the remark this answers is not in the document]
Comment (committedteadrinker) [reply lib/src/neu/verifications/contract-mismatcher.ts]: I updated the method for reusability.

**Comment by committedteadrinker** [comment]
Agreed with you @alephao I think we should revisit the structure of that later on. Though if some methods are returning `Result<>` I think the others should do too? It's to keep consistency across the whole file.

**Review by lfportal — APPROVED** [review]

**Comment by alephao** [comment]
@committedteadrinker I know you're just following the current structure, we should definitely keep consistency. I just wanted to highlight the problem, but it's not really specific to this PR.

**Review by alephao — APPROVED** [review]

**Pull request merged**